tt_metal/impl/trace/trace_buffer.cpp

```diff
@@ -0,0 +1,16 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "trace_buffer.hpp"
+#include "tt_metal/impl/device/device.hpp"
+
+namespace tt::tt_metal {
+
+TraceBuffer::TraceBuffer(std::shared_ptr<detail::TraceDescriptor> desc, std::shared_ptr<Buffer> buffer) : desc(desc), buffer(buffer) {}
+
+TraceBuffer::~TraceBuffer() {
+    this->buffer->device()->trace_buffers_size -= this->buffer->size();
+}
```

Comment:
Need to check it's not nullptr?

Reply:
Device shouldn't be a nullptr after the device_pool changes but I'll add a check for safety. Buffer can't be a nullptr, since its a shared ptr, and will live as long as the TraceBuffer lives

Reply:
nullptr check was for buffer because we create empty tracebuffers right?